### ISSUE TITLE
Allow multiple listeners on same port in Gateway API provider

### ISSUE THREAD

**Issue by TeoGoddet** [traefik/traefik#9026]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

Concern K8SGatewayAPI provider: 
Sould be possible to have multiple HTTP/HTTPS listeners on the same port (the goal of a reverse proxy :)
Now this is forbidden by this line : https://github.com/traefik/traefik/blob/1048348ae65497feb4a05e090f0ac910586c7996/pkg/provider/kubernetes/gateway/kubernetes.go#L343

Am I right ?

**Comment by TeoGoddet** [comment]
The config I would like to be working : 
```
  listeners:
  - allowedRoutes:
      namespaces:
        from: All
    hostname: hello.agepoly.ch
    name: hello.agepoly.ch
    port: 8443
    protocol: HTTPS
  - allowedRoutes:
      namespaces:
        from: All
    hostname: agevote-api.agepoly.ch
    name: agevote-api.agepoly.ch
    port: 8443
    protocol: HTTPS
```

**Comment by kevinpollet** [comment]
Hello @TeoGoddet,

Thanks for your interest in Traefik,

> Am I right ?

You are right, as explained in the [specification](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Gateway), each listener in a Gateway must have a unique combination of Hostname, Port, and Protocol.

So the use case you are trying to achieve is valid.

**Comment by TeoGoddet** [comment]
At first sight : 
```
if _, ok := allocatedPort[listener.Port]; ok {
			listenerStatuses[i].Conditions = append(listenerStatuses[i].Conditions, metav1.Condition{
				Type:               string(v1alpha2.ListenerConditionDetached),
				Status:             metav1.ConditionTrue,
				LastTransitionTime: metav1.Now(),
				Reason:             string(v1alpha2.ListenerReasonPortUnavailable),
				Message:            fmt.Sprintf("Port %d unavailable", listener.Port),
			})

			continue
		}
```
this should be raised only if : 
a TCP listener want an allocated port (any : HTTP,HTTPS,TCP)
an HTTP/HTTPS listener want a TCP allocated port (TCP)

this may need to be raised when a 
HTTP listener ask for an HTTPS already allocated and vice-versa

**Comment by burner-account** [comment]
In theory, a small patch like this you be enough:

```go

if proto, ok := allocatedPort[listener.Port]; ok {
     oldHTTPS := proto == v1alpha2.HTTPSProtocolType
     newHTTPS :=  listener.Protocol == v1alpha2.HTTPSProtocolType
     if !oldHTTPS || !newHTTPS  {
        listenerStatuses[i].Conditions = append(listenerStatuses[i].Conditions, metav1.Condition{
	    Type:               string(v1alpha2.ListenerConditionDetached),
	    Status:             metav1.ConditionTrue,
	    LastTransitionTime: metav1.Now(),
	    Reason:             string(v1alpha2.ListenerReasonPortUnavailable),
	    Message:            fmt.Sprintf("Port %d unavailable", listener.Port),
        })
	continue
    }
}
```
This would retain the original behavior except HTTPS listeners.

**Comment by burner-account** [comment]
> Last thing is what happen if you define exactly the same listener twice (probably routes add together and traefik priority rules apply, it would mean we need to check what the spec tells about priorities and maybe implement some things)

Only thing known to me about weighing are  backendRefs in HTTPRoute - that is where you do A/B tests according to Gateway API.
I would currently restrict this solution to HTTPS protocol and will try to find a cheap way to ensure that listener hostnames differ.

**Comment by burner-account** [comment]
Ok, now there is a PR (#9107 ) which address the following topics:
- different protocols cannot share a port
- multiple listeners are "up to spec" as long as the (protocol, hostname, port) tuple is unique

As there might be a discussion on which protocols multiple listeners make sense, it is configurable for which protocols mult-listeners are enabled. Currently only the HTTPS protocol is enabled for multi-listeners.

**Comment by traefiker** [comment]
Closed by #9107.
